### PR TITLE
Metric exemplar

### LINKED ISSUE #1576: Add Trace Exemplar Attachments to Existing Metrics
**Problem**
Exemplars allow representative traces to be selected from metric data. Opencensus allows a spancontext to be added as an attachment when recording metrics, see https://pkg.go.dev/go.opencensus.io@v0.22.4/metric/metricdata. When recording metrics we should attach the current span context whenever the span is being sampled.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Event consumers and producers.

**Exit Criteria**
Exemplar functionality manually verified in cloud monitoring.

**Time Estimate (optional):** 
3-5 days.

### PR DESCRIPTION
**Pull request by zhongduo**
Fixes #1576 

## Proposed Changes

- Add exemplar attachment to `ReportEventDispatchTime` in `delivery_reporter.go`.
- Add exemplar attachment to `reportEventProcessingTime ` in `delivery_reporter.go`.
- Add exemplar attachment to `ReportEventCount` in `ingress_reporter.go`.

**Release Note**

```release-note
- 🎁 Add exemplar traces to dispatch time and processing time, the exemplar can be viewed in Stackdriver dashboard in heatmap view.
```

**Docs**

### CODE REVIEW

**Comment by zhongduo** [comment]
Several questions about this PR:

1. Do we need test here?
2. Do we want to keep the change in the event count, though `count` view does not support exemplar view in Stackdriver? Note that other exporter might have different implementation.

File: pkg/metrics/delivery_reporter.go [hunk 1]
```diff
@@ -188,3 +192,17 @@ func getStartDeliveryProcessingTime(ctx context.Context) (time.Time, error) {
 	}
 	return time.Time{}, fmt.Errorf("missing or invalid start time: %v", v)
 }
+
+// Get attachment for exemplar trace.
+func getSpanCtxAttachment(ctx context.Context) metricdata.Attachments {
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/metrics/delivery_reporter.go]: Fixed as suggested.

File: pkg/metrics/delivery_reporter.go [hunk 2]
```diff
@@ -188,3 +192,17 @@ func getStartDeliveryProcessingTime(ctx context.Context) (time.Time, error) {
 	}
 	return time.Time{}, fmt.Errorf("missing or invalid start time: %v", v)
 }
+
+// Get attachment for exemplar trace.
```
[the remark this answers is not in the document]
Comment (zhongduo) [reply pkg/metrics/delivery_reporter.go]: Fixed as suggested.

File: pkg/metrics/delivery_reporter.go [hunk 3]
```diff
@@ -188,3 +192,15 @@ func getStartDeliveryProcessingTime(ctx context.Context) (time.Time, error) {
 	}
 	return time.Time{}, fmt.Errorf("missing or invalid start time: %v", v)
 }
+
+// getSpanContextAttachments gets the attachment for exemplar trace.
+func getSpanContextAttachments(ctx context.Context) metricdata.Attachments {
+	attachments := map[string]interface{}{}
+	span := trace.FromContext(ctx)
+	if span != nil {
```
Comment (Harwayne) [inline pkg/metrics/delivery_reporter.go:200]:
```suggestion
	if span := trace.FromContext(ctx); span != nil {
```

**Comment by ian-mi** [comment]
This looks good! I think we may want to expand our metrics e2e test to check for the presence of the trace exemplar when querying cloud monitoring.

**Comment by zhongduo** [comment]
Should we open another issue for that?

**Pull request merged**